File: PRACTICAS/CAMBIAR_RUBY/RUBY2/Civitas/lib/controlador.rb
module Civitas
  class Controlador
    def initialize (_juego, _vista)
      @juego = _juego
      @vista = _vista
    end
    def juega
      @vista.set_civitas_juego(@juego)
      while (!@juego.final_del_juego)
        @vista.actualizar_vista
        @vista.pausa
        operacion = @juego.siguiente_paso
        @vista.mostrar_siguiente_operacion(operacion)
        if (operacion != Operaciones_Juego::PASAR_TURNO)
          @vista.mostrar_eventos
        end
        if (!@juego.final_del_juego)
          if operacion == Operaciones_Juego::COMPRAR

          if (!@vista.juego_model.get_casilla_actual.titulo_propiedad.tiene_propietario && @vista.juego_model.get_casilla_actual.tipo == TipoCasilla::CALLE)
            compra = @vista.comprar

            if compra == Respuestas::SI
              @juego.comprar
              @juego.siguiente_paso_completado(operacion)
            else
              juego.siguiente_paso_completado(operacion)
            end
          else
            if @vista.juego_model.get_casilla_actual.tipo == TipoCasilla::CALLE
              puts "No puedes comprar la propiedad #{@vista.get_casilla_actual.titulo_propiedad.nombre} porque ya tiene propietario.\n"
            else
              puts "\nNo puedes comprar esta casilla porque no es una calle.\n"
            end
            @juego.siguiente_paso_completado(operacion)
          end

        elsif operacion == Operaciones_Juego::GESTIONAR

          lista_gestiones_inmobiliarias = [GestionesInmobiliarias::VENDER, GestionesInmobiliarias::HIPOTECAR,
                                           GestionesInmobiliarias::CANCELAR_HIPOTECA, GestionesInmobiliarias::CONSTRUIR_CASA,
                                           GestionesInmobiliarias::CONSTRUIR_HOTEL,GestionesInmobiliarias::TERMINAR]


          @vista.gestionar
          gestion = @vista.gestion
          propiedad = @vista.propiedad

          g_inmobiliaria = lista_gestiones_inmobiliarias[gestion]
          o_inmobiliaria = GestionesInmobiliarias.new(g_inmobiliaria, propiedad)

          if o_inmobiliaria == GestionesInmobiliarias::VENDER
            @juego.vender(propiedad)
          elsif o_inmobiliaria == GestionesInmobiliarias::HIPOTECAR
            @juego.hipotecar(propiedad)
          elsif o_inmobiliaria == GestionesInmobiliarias::CANCELAR_HIPOTECA
            @juego.cancelar_hipoteca(propiedad)
          elsif o_inmobiliaria == GestionesInmobiliarias::CONSTRUIR_CASA
            @juego.construir_casa(propiedad)
          elsif o_inmobiliaria == GestionesInmobiliarias::CONSTRUIR_HOTEL
            @juego.construir_hotel(propiedad)
          elsif o_inmobiliaria== GestionesInmobiliarias::TERMINAR
            @juego.siguiente_paso_completado(Operaciones_Juego::GESTIONAR)
          end


        elsif operacion == Operaciones_Juego::SALIR_CARCEL

          salida = @vista.salir_carcel

          if salida == SalidasCarcel::PAGANDO
            @vista.juego_model.salir_carcel_pagando
          else

            @vista.juego_model.salir_carcel_tirando
          end

          @vista.juego_model.siguiente_paso_completado(Operaciones_Juego::SALIR_CARCEL)

        end
        end
      end
      ranking = @juego.ranking
      puts ranking
    end
  end
end

File: PRACTICAS/practica3/Civitas_P1/src/JuegoTexto/Controlador.java

/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package JuegoTexto;

import civitas_p1.CivitasJuego;
import civitas_p1.GestionesInmobiliarias;
import civitas_p1.Jugador;
import civitas_p1.OperacionInmobiliaria;
import civitas_p1.OperacionesJuego;
import civitas_p1.Respuestas;
import civitas_p1.SalidasCarcel;
import civitas_p1.TipoCasilla;
import java.util.ArrayList;

/**
 *
 * @author angelsolano
 */
public class Controlador {

    private CivitasJuego juego;
    private VistaTextual vista;
    
    public Controlador(CivitasJuego juego, VistaTextual vista){
        
        this.juego = juego;
        this.vista = vista;
        
    }
    
    public void juega(){
        
        vista.setCivitasJuego(juego);

        while(!juego.finalDelJuego()){
            
            int cnt = 0;
                       
            if (cnt > 0)vista.actualizarVista();
            
            vista.pausa();
            OperacionesJuego operacion = juego.siguientePaso();
            vista.mostrarSiguienteOperacion(operacion);
            if (operacion != OperacionesJuego.PASAR_TURNO)
                vista.mostrarEventos();

            if (!juego.finalDelJuego()){

                if (operacion == OperacionesJuego.COMPRAR){
                    
                    if(!vista.juegoModel.getCasillaActual().getTituloPropiedad().tienePropietario() && vista.juegoModel.getCasillaActual().getTipo() == TipoCasilla.CALLE){
                        Respuestas compra = vista.comprar();
                        
                        if (compra == Respuestas.values()[1]){
                            juego.comprar();
                            juego.siguientePasoCompletado(operacion);
                        }else
                            juego.siguientePasoCompletado(operacion);
                     
                    }else{
                        
                        if (vista.juegoModel.getCasillaActual().getTipo() == TipoCasilla.CALLE)
                            System.out.println("No puedes comprar la propiedad " +vista.juegoModel.getCasillaActual().getTituloPropiedad().getNombre()+" porque ya tiene propietario.");
                        else
                            System.out.println("\nNo puedes comprar esta casilla porque no es una calle.\n");
                        juego.siguientePasoCompletado(operacion);
                    }    
                    

                }
                else if (operacion == OperacionesJuego.GESTIONAR){
                    vista.gestionar();
                    int gestion = vista.getGestion();
                    int propiedad = vista.getPropiedad();
                    GestionesInmobiliarias g_inmobiliaria = GestionesInmobiliarias.values()[gestion];
                    OperacionInmobiliaria o_inmobiliaria= new OperacionInmobiliaria(g_inmobiliaria, propiedad);

                    if (o_inmobiliaria.getGestion() == GestionesInmobiliarias.values()[0]){
                        juego.vender(propiedad);
                    }
                    else if (o_inmobiliaria.getGestion() == GestionesInmobiliarias.values()[1]){
                        juego.hipotecar(propiedad);
                    }
                    else if (o_inmobiliaria.getGestion() == GestionesInmobiliarias.values()[2]){
                        juego.cancelarHipoteca(propiedad);
                    }
                    else if (o_inmobiliaria.getGestion() == GestionesInmobiliarias.values()[3]){
                        juego.construirCasa(propiedad);
                    }
                    else if (o_inmobiliaria.getGestion() == GestionesInmobiliarias.values()[4]){
                        juego.construirHotel(propiedad);
                    }
                    else if (o_inmobiliaria.getGestion() == GestionesInmobiliarias.values()[5]){
                        juego.siguientePasoCompletado(operacion);
                    }

                }
                else if (operacion == OperacionesJuego.SALIR_CARCEL){
                    SalidasCarcel salida = vista.salirCarcel();

                    if (salida == SalidasCarcel.values()[0])
                        juego.salirCarcelPagando();
                    else
                        juego.salirCarcelTirando();

                    juego.siguientePasoCompletado(operacion);
                }
            }  
            
            cnt ++;
        }
    
    ArrayList<Jugador> top_players = juego.ranking();
    
    }
}
